OCPNODE-1499: Add support for must-gather collection script

jrvaldes opened the pull request:
This change introduces a specific must-gather collection script for this operator within the operator image itself, so that it can be used directly in the Openshift' must-gather process.

<!-- Thank you for contributing!

     Read more about how you can contribute in our contribution guide:
     https://github.com/kedacore/keda/blob/master/CONTRIBUTING.md
-->


### Checklist

- [ x] Commits are signed with Developer Certificate of Origin (DCO)

jrvaldes commented on Dockerfile:
```diff
@@ -23,17 +23,26 @@ COPY main.go main.go
+# Use Red Hat Universal Base Image Micro as minimal base image to package the manager binary.
+# See https://catalog.redhat.com/software/containers/ubi8/ubi-micro/5ff3f50a831939b08d1b832a
+# TODO: Use authenticated registry: registry.redhat.io
+FROM registry.access.redhat.com/ubi8/ubi-micro:8.7-6
```
Should the authenticated registry be used instead?

joelsmith replied:
We use currently use registry.redhat.io/ubi8/ubi-minimal:latest
see:
https://gitlab.cee.redhat.com/custom-metrics-autoscaler/cpaas-midstream/-/blob/ocma-2-rhel-8/distgit/containers/custom-metrics-autoscaler-operator/Dockerfile.in

jrvaldes replied:
Addressed in https://github.com/openshift/custom-metrics-autoscaler-operator/pull/21/commits/f266c048f1a03f55f5459fec2b5e02f7e4edf428

jrvaldes commented on Dockerfile:
```diff
@@ -23,17 +23,26 @@ COPY main.go main.go
 
-# Use distroless as minimal base image to package the manager binary
-# Refer to https://github.com/GoogleContainerTools/distroless for more details
-FROM gcr.io/distroless/static:nonroot
```
I'm proposing `ubi8/ubi-micro` as a replacement for `distroless` as a base image for the operator's manager library, so that `/bin/bash` is available to run the must-gather collection script. 

See https://www.redhat.com/en/blog/why-distroless-containers-arent-security-solution-you-think-they-are

joelsmith replied:
This is an upstream Dockerfile. For our build we use this Dockerfile:
https://gitlab.cee.redhat.com/custom-metrics-autoscaler/cpaas-midstream/-/blob/ocma-2-rhel-8/distgit/containers/custom-metrics-autoscaler-operator/Dockerfile.in

jrvaldes commented on must-gather/README.md:
````diff
@@ -0,0 +1,117 @@
+```sh
+oc adm must-gather \
+  --image-stream=openshift/must-gather \
+  --image=registry.redhat.io/custom-metrics-autoscaler-tech-preview/custom-metrics-autoscaler-rhel8-operator
````
Does it make sense to push the operator image to Quay? So that the must-gather command may look like this:

```
oc adm must-gather \
  --image-stream=openshift/must-gather \
  --image=quay.io/openshift/origin-custom-metrics-autoscaler-operator
```

joelsmith replied:
I prefer that we use the image in cluster for the docs. That way it will work for any cluster with CMA available, including disconnected ones (assuming they have mirrored the CMA images).

So my preference would be either a two-step:
```
IMAGE="$(oc get packagemanifests -n openshift-marketplace openshift-custom-metrics-autoscaler-operator -o jsonpath='{.status.channels[?(@.name=="stable")].currentCSVDesc.annotations.containerImage}')"
oc adm must-gather --image="$IMAGE"
```

or a one-step:
```
oc adm must-gather --image="$(oc get packagemanifests -n openshift-marketplace openshift-custom-metrics-autoscaler-operator -o jsonpath='{.status.channels[?(@.name=="stable")].currentCSVDesc.annotations.containerImage}')"
```
```

jrvaldes replied:
Agreed, I lean towards leaving only the [first](https://github.com/openshift/custom-metrics-autoscaler-operator/pull/21/files#diff-f22cb59a526d4b8f524cafa3abb53f4ced4afc9b32ef3b3a420ed87e4b870d2eR8) alternative using the container image of the existing deployment.

joelsmith replied:
The issue there is that the deployment won't exist if the operator failed to install. The packagemanifests version should work in any situation as long as the catalog has the entry.

jrvaldes replied:
True.

jrvaldes replied:
> IMAGE="$(oc get packagemanifests -n openshift-marketplace openshift-custom-metrics-autoscaler-operator -o jsonpath='{.status.channels[?(@.name=="stable")].currentCSVDesc.annotations.containerImage}')"
oc adm must-gather --image="$IMAGE"

@joelsmith  As per above, can we assume that using the `stable` channel is always safe?

joelsmith replied:
Yes. It's the only channel we have 😄

joelsmith commented:
As this repo is a fork of upstream https://github.com/kedacore/keda-olm-operator, we want our OpenShift-specific changes to be minimal so that we don't have to worry as much about merge conflicts, etc. when we do our periodic upstream rebase operations.

At a miniumum, the `cma-verify-history` will require that your commit(s) be prefixed with `UPSTREAM: <carry>:` so that we don't accidentally drop them on the next rebase.

The dockerfile changes should be reverted, and re-applied to our gitlab CPAAS midstream repo (https://gitlab.cee.redhat.com/custom-metrics-autoscaler/cpaas-midstream/-/blob/ocma-2-rhel-8/distgit/containers/custom-metrics-autoscaler-operator/Dockerfile.in)

I am having a hard time deciding whether it's worth updating the root README.md given the doctoc issue you identified. Either way, we'll certainly ask @mburke5678 to change the OpenShift product docs to include this info. What do you think?

joelsmith commented:
This is looking good. I'm ready to give `lgtm` once the fixup commits are squashed.

jrvaldes commented:
> This is looking good. I'm ready to give `lgtm` once the fixup commits are squashed.

Thanks @joelsmith . I squashed the commits and updated the PR. PTAL 

The dockerfile changes are captured in the [midstream MR](https://gitlab.cee.redhat.com/custom-metrics-autoscaler/cpaas-midstream/-/merge_requests/80).

jrvaldes commented:
/CC @mburke5678 

For docs updates.

jrvaldes commented:
Pushed an update to generate the must-gather version file [based](https://github.com/openshift/custom-metrics-autoscaler-operator/commit/64f3fecd86f5d9404156d56e5bba03606c8da603#diff-e5c2a69519b4a8fddd46d182316399558c9fd41434650a1196c7a9beaa14ada0R39) on the information provided by the packagemanifest.

jrvaldes commented on must-gather/collection-scripts/gather:
```diff
@@ -0,0 +1,66 @@
+CMA_SUB_NAME="openshift-custom-metrics-autoscaler-operator"
+
+# find namespace
+CMA_NS=$(oc get subs -A -o template --template '{{range .items}}{{if eq .spec.name "'"${CMA_SUB_NAME}"'"}}{{.metadata.namespace}}{{end}}{{end}}')
```
@joelsmith  I'm leaning on the subscription to find the namespace where the operator was installed. If the CMA subscription is absent, this logic assumes the operator is not present; hence, there is nothing to collect. Thoughts?

Should the logic entertain other cases, the operator was deployed directly (without OLM), etc.?

joelsmith replied:
I think this is fine, especially since this is a support tool, and our supported use case is installation via OLM.

joelsmith commented:
/lgtm
/approve
Thanks!

jrvaldes commented:
CMA is single stream, hence no backports are needed.
